language/diem-framework/src/main.rs

```diff
@@ -104,15 +104,29 @@ fn main() {
                 let mut module_path = PathBuf::from(COMPILED_OUTPUT_PATH);
                 module_path.push(COMPILED_STDLIB_DIR);
                 let new_modules = build_stdlib();
+                let new_modules_by_id: BTreeMap<_, _> = new_modules
+                    .values()
+                    .map(|module| (module.self_id(), module))
+                    .collect();
 
                 let mut is_linking_layout_compatible = true;
                 if !no_check_linking_layout_compatibility {
-                    for module in new_modules.values() {
+                    for (new_module_id, module) in &new_modules_by_id {
                         // extract new linking/layout API and check compatibility with old
-                        let new_module_id = module.self_id();
                         if let Some(old_api) = old_module_apis.get(&new_module_id) {
+                            let old_friends: Vec<_> = old_api
```

Comment:
Is it possible to put this logic inside `Compatibility::check`? It would be great if the correctness of this function didn't depend on the caller computing the `old_friends` correctly.

Reply:
I think it is going to be difficult as `Compatibility::check` does not know how to retrieve modules by `module_id`. We could pass an oracle `Fn(&ModuleId) -> &CompiledModule` to `Compatibility::check`, but it is still the caller's job to give the oracle.

Reply:
Ah, I see the problem... 

I guess it's not clear that you should be allowed to remove `friend` functions. We don't allow this for any other kind of exposed functions, and from a policy perspective you could imagine a `friend` module wanting to sign off on being removed even if it's not a current caller (perhaps it plans to add a call site in the future).

My inclination would be to remove this ability for now, but perhaps reconsider it once we add simultaneous publishing/upgrades of modules (which I think may present some similar problems). Thoughts?

Reply:
I am fine either way. We could either change the behavior to now to disallow `friend` function removals, or update this function once simultaneous publishing is ready.

Reply:
In that case, perhaps disallow for now? In the future, if we do allow this, I think we can reduce the reliance on the caller by validating that `old_module.friends == old_friends.iter().map(|m| m.self_id())` (or similar).